File: LCA/constant_time/python/lca_to_rmq.py
from typing import List
from math import log2
from math import ceil
class SparseTable:
    def __init__(self,arr:List[int]):
        self.n = len(arr)
        self.l = int(log2(self.n))
        self.rmq = [[-float('inf')]*(self.l+1) for _ in range(self.n)]
        for i in range(self.n):
            self.rmq[i][0] = arr[i]
        self.build()
    def f(self,a:int,b:int):
        return min(a,b)
    def build(self):
        j = 1
        while j<=self.l:
            i = 0
            while i+(1<<j)<=self.n:
                self.rmq[i][j] = self.f(self.rmq[i][j-1],self.rmq[i+(1<<(j-1))][j-1])
                i+=1
            j+=1
    def query(self,l:int,r:int):
        j = int(log2(r-l+1))
        return self.f(self.rmq[l][j],self.rmq[r-(1<<j)+1][j])
class LCA:
    def __init__(self,adj:List[List[int]],n:int,root:int=0):
        self.adj = adj 
        self.tin = [0]*n
        self.tour = []
        self.time_to_node = [0]*(2*n)
        self.dfs(root,root)
        self.rmq = SparseTable(self.tour)
    def dfs(self,u:int,p:int)->None:
        self.tin[u] = len(self.tour)
        self.time_to_node[self.tin[u]] = u
        self.tour.append(self.tin[u])
        for v in self.adj[u]:
            if v!=p:
                self.dfs(v,u)
                self.tour.append(self.tin[u])
    def lca(self,u:int,v:int)->int:
        if self.tin[u]>self.tin[v]: u,v = v,u 
        return self.time_to_node[self.rmq.query(self.tin[u],self.tin[v])]
if __name__=="__main__":
    adj = [[4],[4],[3],[4,2],[3,1,0]]
    lca = LCA(adj,5,3)
    # Expected value = 4
    print(lca.lca(0,1))

File: LCA/constant_time/cpp/lca_to_rmq.cpp
#include<bits/stdc++.h>
#define INF 1e9
using namespace std;
struct SparseTable{
    vector<vector<int>> rmq;
    int n,l;
    SparseTable(){}
    SparseTable(vector<int> &arr){
        n = (int)arr.size();
        l = __lg(n);
        rmq.assign(n,vector<int>(l+1,-INF));
        for(int i=0;i<n;i++)rmq[i][0] = arr[i];
        build();
    }
    int f(int &a,int &b){
        return min(a,b);
    }
    void build(){
        for(int j=1;j<=l;j++)
            for(int i=0;i+(1<<j)<=n;i++)
                rmq[i][j] = f(rmq[i][j-1],rmq[i+(1<<(j-1))][j-1]);
    }
    int query(int l,int r){
        int j = __lg(r-l+1);
        return f(rmq[l][j],rmq[r-(1<<j)+1][j]);
    }
};
struct LCA{
    vector<vector<int>> adj;
    vector<int> tour,tin,time_to_node;
    SparseTable rmq;
    LCA(vector<vector<int>> &adj,int n,int root=0):adj(adj){
        tin.resize(n);
        time_to_node.resize(2*n);
        dfs(root,root);
        rmq = SparseTable(tour);
    }
    void dfs(int u,int p){
        tin[u] = (int)tour.size();
        time_to_node[tin[u]] = u;
        tour.push_back(tin[u]);
        for(int v:adj[u])
            if(v!=p)dfs(v,u),tour.push_back(tin[u]); 
    }
    int lca(int u,int v){
        if(tin[u]>tin[v])swap(u,v);
        return time_to_node[rmq.query(tin[u],tin[v])];
    }
};
int main(){
    vector<vector<int>> adj({{4},{4},{3},{4,2},{3,1,0}});
    LCA lca = LCA(adj,5,3);
    // Expected value = 4
    cout<<lca.lca(0,1)<<"\n";
    return 0;
}